angular-protractor: Incorrect generic type arg in protractor.ElementArrayFinder.map

mapFn does not always return T as madated by the declaration below:
	
	map<T>(mapFn: (element: ElementFinder, index: number) => T): webdriver.promise.Promise<T[]>;

Let's imagine  T is an instance of *elementInfo*:
	
	interface elementInfo {
		index: number;
		text?: string;
	}

mapFn will only return T in cases where no promises are included in the object it returns as below:
	
	(elm: protractor.ElementFinder, index: number) => {
		return {
			index: index //number
		};
	}

mapFn will *not* return T in cases where promises are included in the object it returns as below:

	(elm: protractor.ElementFinder, index: number) => {
		return {
			index: index, //number
			text: elm.getText() //webdriver.promise.Promise<string>
		};
	}

*map* will still return elementInfo in both cases. Only *mapFn* is affected.

**Example 1** - This should be perfectly valid and is so at runtime but fails to compile:
        
	interface elementInfo {
		index: number;
		text: string;
	}

	interface elementInfoUnresolved {
		index: number;
		text: webdriver.promise.Promise<string>;
	}

	function getTargetIndex(): webdriver.promise.Promise<number> {

		var targetIndex: number = null;

		return element.all(by.css('#table tbody tr td')).map<elementInfo>((elm: protractor.ElementFinder, index: number): elementInfoUnresolved => {
			return {
				index: index,
				text: elm.getText()
			};
		}).then((items: elementInfo[])=>{
			items.forEach((item: any)=>{
				if (item.text.indexOf('test') != -1) {
					targetIndex = item.index;
				}
			});
			return targetIndex;
		});

	}


**Result:**
	
	error TS2345: Argument of type '(elm: ElementFinder  
	, index: number) => elementInfoUnresolved' is not assignable to parameter of type '(element: Elemen  
	tFinder, index: number) => elementInfo'.                                                             
	>>   Type 'elementInfoUnresolved' is not assignable to type 'elementInfo'.                           
	>>     Types of property 'text' are incompatible.                                                    
	>>       Type 'Promise<string>' is not assignable to type 'string'.

The function below will always return *elementInfoUnresolved* at runtime but acording to the declaration it must return *elementInfo*.
        
	(elm: protractor.ElementFinder, index: number) => {
		return {
			index: index, //number
			text: elm.getText() //webdriver.promise.Promise<string>
		};
	}

The source of confusion is that the function above is iterated over several times returning *elementInfoUnresolved* each time. The getText() promises in each elementInfoUnresolved object are resolved to form elementInfo objects *after* this and an array of the resulting *elementInfo* objects is passed in to the next link in the chain:
        
	.then((items: elementInfo[])=>{
		items.forEach((item: any)=>{
			if (item.text.indexOf('test') != -1) {
				targetIndex = item.index;
			}
		});
		return targetIndex;
	});


**Example 2** - Using any as a workaround is the only way to get it to compile:

	interface elementInfo {
		index: number;
		text: string;
	}

	interface elementInfoUnresolved {
		index: number;
		text: webdriver.promise.Promise<string>;
	}

	function getTargetIndex(): webdriver.promise.Promise<number> {

		var targetIndex: number = null;

		return element.all(by.css('#table tbody tr td')).map<elementInfo>((elm: protractor.ElementFinder, index: number): any=> {
			return {
				index: index, //number
				text: elm.getText() //webdriver.promise.Promise<string>
			};
		}).then((items: elementInfo[])=>{
			items.forEach((item: any)=>{
				if (item.text.indexOf('test') != -1) {
					targetIndex = item.index;
				}
			});
			return targetIndex;
		});

	}


**Solution** - Add an additional return type for mapFn:
	
	map<T>(mapFn: (element: ElementFinder, index: number) => T): webdriver.promise.Promise<T[]>;
	map<T, T2>(mapFn: (element: ElementFinder, index: number) => T2): webdriver.promise.Promise<T[]>;

I will submit a pull request shortly

--- a/angular-protractor/angular-protractor-tests.ts
+++ b/angular-protractor/angular-protractor-tests.ts
@@ -406,9 +406,19 @@ function TestElementArrayFinder() {
     elementArrayFinder.each(function(element: protractor.ElementFinder){
         // nothing
     });
+
     stringPromise = elementArrayFinder.map(function(element: protractor.ElementFinder, index: number){
         return 'abc';
-    })
+    });
+
+    stringPromise = elementArrayFinder.map<string>(function(element: protractor.ElementFinder, index: number): string {
+        return 'abc';
+    });
+
+    stringPromise = elementArrayFinder.map<string, webdriver.promise.Promise<string>>(function(element: protractor.ElementFinder, index: number): webdriver.promise.Promise<string> {
+        return element.getText();
+    });
+
     elementArrayFinder = elementArrayFinder.filter(function(element: protractor.ElementFinder, index: number){
         return element.getText().then((text: string) => {
             return text === "foo";

--- a/angular-protractor/angular-protractor.d.ts
+++ b/angular-protractor/angular-protractor.d.ts
@@ -992,6 +992,7 @@ declare module protractor {
         *     of values returned by the map function.
         */
         map<T>(mapFn: (element: ElementFinder, index: number) => T): webdriver.promise.Promise<T[]>;
+        map<T, T2>(mapFn: (element: ElementFinder, index: number) => T2): webdriver.promise.Promise<T[]>;
 
         /**
         * Apply a filter function to each element within the ElementArrayFinder. Returns

